trace_replay/io_tracer.h

```diff
@@ -0,0 +1,129 @@
+  uint64_t len = 0;
+  uint64_t offset = 0;
+
+  IOTraceRecord() {}
+
+  IOTraceRecord(const uint64_t& _access_timestamp, const TraceType& _trace_type,
+                const std::string& _file_operation,
+                const std::string& _io_status, const std::string& _file_name,
+                const uint64_t& _len, const uint64_t& _offset)
+      : access_timestamp(_access_timestamp),
+        trace_type(_trace_type),
+        file_operation(_file_operation),
+        io_status(_io_status),
+        file_name(_file_name),
+        len(_len),
+        offset(_offset) {}
+};
+
+struct IOTraceHeader {
+  uint64_t start_time;
+  uint32_t rocksdb_major_version;
+  uint32_t rocksdb_minor_version;
+};
+
+// IOTraceWriter writes IO operation as a single trace. Each trace will have a
+// timestamp and type, followed by the trace payload.
+class IOTraceWriter {
+ public:
+  IOTraceWriter(Env* env, const TraceOptions& trace_options,
+                std::unique_ptr<TraceWriter>&& trace_writer);
+  ~IOTraceWriter() = default;
+  // No copy and move.
+  IOTraceWriter(const IOTraceWriter&) = delete;
+  IOTraceWriter& operator=(const IOTraceWriter&) = delete;
+  IOTraceWriter(IOTraceWriter&&) = delete;
+  IOTraceWriter& operator=(IOTraceWriter&&) = delete;
+
+  Status WriteIOOp(const IOTraceRecord& record, const Slice& file_operation,
+                   const Slice& io_status, const Slice& file_name);
+
+  // Write a trace header at the beginning, typically on initiating a trace,
+  // with some metadata like a magic number and RocksDB version.
+  Status WriteHeader();
+
+ private:
+  Env* env_;
+  TraceOptions trace_options_;
+  std::unique_ptr<TraceWriter> trace_writer_;
+};
+
+// IOTraceReader helps read the trace file generated by IOTraceWriter.
+class IOTraceReader {
+ public:
+  explicit IOTraceReader(std::unique_ptr<TraceReader>&& reader);
+  ~IOTraceReader() = default;
+  // No copy and move.
+  IOTraceReader(const IOTraceReader&) = delete;
+  IOTraceReader& operator=(const IOTraceReader&) = delete;
+  IOTraceReader(IOTraceReader&&) = delete;
+  IOTraceReader& operator=(IOTraceReader&&) = delete;
+
+  Status ReadHeader(IOTraceHeader* header);
+
+  Status ReadIOOp(IOTraceRecord* record);
+
+ private:
+  std::unique_ptr<TraceReader> trace_reader_;
+};
+
+// An IO tracer. It uses IOTraceWriter to write the access record to the
+// trace file.
+class IOTracer {
+ public:
+  IOTracer();
+  ~IOTracer();
+  // No copy and move.
+  IOTracer(const IOTracer&) = delete;
+  IOTracer& operator=(const IOTracer&) = delete;
+  IOTracer(IOTracer&&) = delete;
+  IOTracer& operator=(IOTracer&&) = delete;
+
+  // Start writing IO operations to the trace_writer.
+  Status StartIOTrace(Env* env, const TraceOptions& trace_options,
+                      std::unique_ptr<TraceWriter>&& trace_writer);
+
+  // Stop writing IO operations to the trace_writer.
+  void EndIOTrace();
+
+  bool is_tracing_enabled() const {
+    return writer_.load(std::memory_order_relaxed);
```

Comment:
Does ```writer_``` need to be an atomic variable? Access to ```writer_``` is already protected by ```trace_writer_mutex_```. ```is_tracing_enabled()``` seems like it would be more frequently called, so it would be better to find a cheaper way to do the check than an atomic variable.

Reply:
@anand1976  `writer_`  is protected by `trace_writer_mutex_ ` in `StartIOTrace` and `EndIOTrace`. So if I change `writer_` to not be atomic, can I access the value of `writer_` without `trace_writer_mutex_ `  in `is_tracing_enabled()`, Or is it fine to do so.

Reply:
Can we just use a ```bool``` to determine if tracing is enabled? ```writer_``` can continue to be an atomic variable, and ```WriteIOOp``` anyway checks ```writer_``` and acquires mutex, so the worst that can happen is a caller calls ```is_tracing_enabled()```, and then calls ```WriteIOOp```, which would simply ignore it if tracing has been disabled in the meantime.